galaxy_ng/app/api/ui/viewsets/tags.py

```diff
@@ -21,3 +30,106 @@ def list(self, request, *args, **kwargs):
+            order = "-" if "-count" in value else ""
+
+            return qs.filter(
+                ansible_collectionversion__ansible_crossrepositorycollectionversionindex__is_highest=True  # noqa: E501
+            ).annotate(count=Count('ansible_collectionversion')).order_by(f"{order}count")
+
+        return super().filter(qs, value)
+
+
+class TagFilter(filterset.FilterSet):
+    sort = TagFilterOrdering(
+        fields=(
+            ("name", "name"),
+            ('count', 'count')
+        ),
+    )
+
+    class Meta:
+        model = Tag
+        fields = {
+            "name": ["exact", "icontains", "contains", "startswith"],
+        }
+
+
+class CollectionsTagsViewSet(
+    api_base.GenericViewSet,
+    mixins.ListModelMixin
+):
+    """
+    ViewSet for collections' tags within the system.
+    """
+    serializer_class = TagSerializer
+    permission_classes = [access_policy.TagsAccessPolicy]
+    versioning_class = versioning.UIVersioning
+    filter_backends = (DjangoFilterBackend,)
+    filterset_class = TagFilter
+
+    queryset = Tag.objects.all()
+
+
+class RolesTagsViewSet(api_base.GenericViewSet):
+    """
+    ViewSet for roles' tags within the system.
+    """
+    queryset = LegacyRole.objects.all()
+    permission_classes = [access_policy.TagsAccessPolicy]
+    versioning_class = versioning.UIVersioning
+    filter_backends = (DjangoFilterBackend,)
+
+    ordering_fields = ["name", "count"]
+    ordering = ["name"]
+    filter_fields = ["exact", "icontains", "contains", "startswith"]
+
+    def _filter_queryset(self, queryset, request):
+        """
+            Custom sorting and filtering,
+            must be performed manually since
+            we are overwriting the queryset with a list of tags.
+        """
+
+        query_params = request.query_params.copy()
+        sort = query_params.get("sort")
+        if sort:
+            query_params.pop("sort")
+
+        # filtering
+        if value := query_params.get("name"):
+            queryset = list(filter(lambda x: x["name"] == value, queryset))
+        elif value := query_params.get("name__contains"):
+            queryset = list(filter(lambda x: value in x["name"], queryset))
+        elif value := query_params.get("name__icontains"):
+            queryset = list(filter(lambda x: value.lower() in x["name"].lower(), queryset))
+        elif value := query_params.get("name__startswith"):
+            queryset = list(filter(lambda x: x["name"].startswith(value), queryset))
+
+        # sorting
+        if sort is not None and sort in ["name", "-name", "count", "-count"]:
+            reverse = True if "-" in sort else False
+            sort_field = sort.replace("-", "")
+            queryset = sorted(queryset, key=lambda x: x[sort_field], reverse=reverse)
+        elif sort is not None:
+            raise serializers.ValidationError(_("Invalid Sort: '{}'").format(sort))
+
+        return queryset
+
+    def list(self, request, *args, **kwargs):
```

Comment:
Is this performant with 40K+ roles?

Reply:
so I did some performance testing on my localhost with locust, synced 5k roles:
![Screenshot from 2023-10-24 13-58-09](https://github.com/ansible/galaxy_ng/assets/19647757/32da3494-83cc-4634-9082-407e59a176ab)


with 100 users concurently it's different story though ![Screenshot from 2023-10-24 13-47-53](https://github.com/ansible/galaxy_ng/assets/19647757/b4210403-9902-44fe-8d6d-d1e607081401)

I'm wondering if caching would be possible and helpful, or to do this we would have to create a role tags model (and probably a model for `full_metadata` as well)?